pkg/daemon/update.go

```diff
@@ -122,6 +122,29 @@ func (dn *Daemon) performPostConfigChangeAction(postConfigChangeActions []string
 	return dn.triggerUpdateWithMachineConfig(state.currentConfig, state.desiredConfig, true)
 }
 
+func setRunningKargsWithCmdline(config *mcfgv1.MachineConfig, requestedKargs []string, cmdline []byte) error {
+	splits := strings.Split(strings.TrimSpace(string(cmdline)), " ")
+	config.Spec.KernelArguments = nil
+	for _, split := range splits {
+		for _, reqKarg := range requestedKargs {
+			if strings.ReplaceAll(reqKarg, "\"", "") == strings.ReplaceAll(split, "\"", "") {
+				config.Spec.KernelArguments = append(config.Spec.KernelArguments, reqKarg)
+				break
+			}
+		}
+	}
+	logSystem("requested kargs: %+v, merged kargs: %+v", requestedKargs, config.Spec.KernelArguments)
+	return nil
+}
+
+func setRunningKargs(config *mcfgv1.MachineConfig, requestedKargs []string) error {
+	b, err := os.ReadFile("/proc/cmdline")
```

Comment:
There's very similar code in `validateKernelArguments`, which parses `rpm-ostree kargs` (which is basically the same thing as `/proc/cmdline` in practice).

But, we also already have a `parseKernelArguments` function here.

(BTW, I do think longer term we need to drop this kernel argument handling out of the MCO; I think https://github.com/containers/bootc/issues/22#issuecomment-1701131496 may be the best path to do it)

Reply:
`parseKernelArguments` parses the kernel arguments from machine config, not from cmdline.
Although it may be used for splitting the cmdline assuming the first and only element of the provided slice is the cmdline.
I am not sure how `validateKernelArguments` can help for comparing the existing kernel arguments with the desired kernel arguments, since it is used for validation.

Reply:
It seems there is attempt in `parseKernelArguments` to parse each of the kernel arguments provided in the machine config.   On the other hand KernelArguments that were not parsed in machine-configs are compared in the function `compareMachineConfig`.

Reply:
@cgwalters what are you suggesting here?

Reply:
Basically I think we shouldn't have one part of the MCO code reading `/proc/cmdline` and another running `rpm-ostree kargs`.  They're not the same thing, but...what we're doing is aiming to be close, and the quotation bug below shows why it's good to only have one path.